lib/node/runner/jsonrpc.go

```diff
@@ -1,55 +1,255 @@
+type DBOpenSnapshot struct{}
+
+type DBOpenSnapshotResult struct {
+	Snapshot string `json:"snapshot"`
+}
+
+type DBReleaseSnapshot struct {
+	Snapshot string `json:"snapshot"`
+}
+
+type DBReleaseSnapshotResult bool
+
+type DBHasArgs struct {
+	Snapshot string `json:"snapshot"`
+	Key      string `json:"key"`
+}
+
 type DBHasResult bool
 
-type DBGetArgs string
+type DBGetArgs struct {
+	Snapshot string `json:"snapshot"`
+	Key      string `json:"key"`
+}
+
 type DBGetResult storage.IterItem
 
 type GetIteratorOptions struct {
-	Reverse bool
-	Cursor  []byte
-	Limit   uint64
+	Reverse bool   `json:"reverse"`
+	Cursor  []byte `json:"cursor"`
+	Limit   uint64 `json:"limit"`
 }
 
 type DBGetIteratorArgs struct {
-	Prefix  string
-	Options GetIteratorOptions
+	Snapshot string             `json:"snapshot"`
+	Prefix   string             `json:"prefix"`
+	Options  GetIteratorOptions `json:"options"`
 }
 
 type DBGetIteratorResult struct {
-	Limit uint64
-	Items []storage.IterItem
+	Limit uint64             `json:"limit"`
+	Items []storage.IterItem `json:"items"`
 }
 
 type jsonrpcDBApp struct {
-	st *storage.LevelDBBackend
+	st        *storage.LevelDBBackend
+	snapshots *expireSnapshots
+}
+
+type expireSnapshots struct {
+	sync.RWMutex
+	st           *storage.LevelDBBackend
+	interval     time.Duration
+	maxSnapshots uint64
+	ticker       *time.Ticker
+	snapshots    *syncmap.Map
+	expires      *syncmap.Map
+}
+
+func newExpireSnapshots(st *storage.LevelDBBackend, interval time.Duration, maxSnapshots uint64) *expireSnapshots {
+	return &expireSnapshots{
+		st:           st,
+		interval:     interval,
+		maxSnapshots: maxSnapshots,
+		ticker:       time.NewTicker(time.Second * 10),
+		snapshots:    &syncmap.Map{},
+		expires:      &syncmap.Map{},
+	}
+}
+
+func (j *expireSnapshots) len() (l int) {
+	j.snapshots.Range(func(_, _ interface{}) bool {
+		l++
+		return true
+	})
+
+	return
+}
+
+func (j *expireSnapshots) newSnapshot() (string, *storage.LevelDBBackend, error) {
+	if j.len() >= int(j.maxSnapshots) {
+		return "", nil, errors.SnapshotLimitReached
+	}
+
+	j.Lock()
+	defer j.Unlock()
+
+	st, err := j.st.OpenSnapshot()
+	if err != nil {
+		return "", nil, err
+	}
+
+	key := common.GetUniqueIDFromUUID()
+	j.snapshots.Store(key, st)
+	j.updateExpire(key)
+
+	log.Debug("new snapshot created", "key", key, "current", j.len())
+	return key, st, nil
+}
+
+func (j *expireSnapshots) snapshot(key string) (*storage.LevelDBBackend, bool) {
+	j.RLock()
+	defer j.RUnlock()
+
+	s, ok := j.snapshots.Load(key)
+	if !ok {
+		return nil, false
+	}
+
+	j.updateExpire(key)
+	return s.(*storage.LevelDBBackend), true
+}
+
+func (j *expireSnapshots) expire(key string) bool {
+	go func() {
+		log.Debug("snapshot expired", "key", key, "current", j.len())
+	}()
+	return j.release_(key)
+}
+
+func (j *expireSnapshots) release(key string) bool {
+	go func() {
+		log.Debug("snapshot released", "key", key, "current", j.len())
+	}()
+	return j.release_(key)
+}
```

Comment:
Bodies of expireSnapshots.`expire()` and expireSnapshots.`release()` are almost same except log messages. What is your intention?

Reply:
@Charleslee522 I wanted to split the expire and release action explicitly.